src/common/utils/FocusManager.ts

```diff
@@ -0,0 +1,387 @@
+/**
+* FocusManager.ts
+*
+* Copyright (c) Microsoft Corporation. All rights reserved.
+* Licensed under the MIT license.
+*
+* Manages focusable elements for better keyboard navigation.
+*/
+
+import React = require('react');
+import PropTypes = require('prop-types');
+
+let _lastComponentId: number = 0;
+
+export interface StoredFocusableComponent {
+    id: string;
+    numericId: number;
+    component: React.Component<any, any>;
+    onFocus: () => void;
+    restricted: boolean;
+    limitedCount: number;
+    origTabIndex?: number;
+    origAriaHidden?: string;
+    removed?: boolean;
+    callbacks?: FocusableComponentStateCallback[];
+}
+
+export interface OriginalAttributeValues {
+    tabIndex: number|undefined;
+    ariaHidden: string|undefined;
+}
+
+export type FocusableComponentStateCallback = (restrictedOrLimited: boolean) => void;
+
+export abstract class FocusManager {
+    private static _rootFocusManager: FocusManager;
+
+    private static _restrictionStack: FocusManager[] = [];
+    private static _currentRestrictionOwner: FocusManager|undefined;
+    private static _restoreRestrictionTimer: number|undefined;
+    private static _pendingPrevFocusedComponent: StoredFocusableComponent|undefined;
+    protected static _currentFocusedComponent: StoredFocusableComponent|undefined;
+    protected static _allFocusableComponents: { [id: string]: StoredFocusableComponent } = {};
+    protected static _skipFocusCheck = false;
+    protected static _resetFocusTimer: number | undefined;
+
+    private _parent: FocusManager|undefined;
+    private _isFocusLimited: boolean;
+    private _prevFocusedComponent: StoredFocusableComponent|undefined;
+    private _myFocusableComponentIds: { [id: string]: boolean } = {};
+
+    constructor(parent: FocusManager|undefined) {
+        if (parent) {
+            this._parent = parent;
+        } else if (FocusManager._rootFocusManager) {
+            console.error('FocusManager: root is already set');
+        } else {
+            FocusManager._rootFocusManager = this;
+        }
+    }
+
+    protected abstract /* static */ addFocusListenerOnComponent(component: React.Component<any, any>, onFocus: () => void): void;
+    protected abstract /* static */ removeFocusListenerFromComponent(component: React.Component<any, any>, onFocus: () => void): void;
+    protected abstract /* static */ focusComponent(component: React.Component<any, any>): boolean;
+
+    // protected abstract /* static */ focusFirst(last?: boolean): void;
```

Comment:
Why is static commented and why is focusFirst commented?
If there is a reason, I guess we need to remove the commented code completely.

Reply:
Yeah, right, a leftover. Each flavor has its own focusFirst called from the corresponding resetFocus, etc.

---
In reply to: [159244518](https://github.com/Microsoft/reactxp/pull/404#discussion_r159244518) [](ancestors = 159244518)